### ISSUE TITLE
[FR] Add ability to create benchmark arguments via cartesian product

### ISSUE THREAD

**Issue by chriswasser** [google/benchmark#1028]
**Is your feature request related to a problem? Please describe.**
Currently, it is only possible to reduce long lists of `->Args({...})` calls via `->Ranges({...})`. While I agree that this interface offers a concise syntax for generating a large number of arguments, I often find myself needing more fine-grained control over the set of generated inputs than `->RangeMultiplier(...)` can provide. This may be due to the requirement for specific values or different range multipliers for different arguments.

**Describe the solution you'd like**
I would like to propose an extension to the public `Benchmark` interface by adding a function `->CartesianProduct({...})` which generates the cartesian product of all the supplied vectors. This would enable fine-grained control over the constructed benchmark arguments without sacrificing on the compact syntax. As a small example:
```cpp
BENCHMARK(BM_SetInsert)
    ->CartesianProduct({{1, 2, 3}, {10}, {6, 12, 24, 36}})
// would be equivalent to
BENCHMARK(BM_SetInsert)
    ->Args({1, 10, 6})
    ->Args({1, 10, 12})
    ->Args({1, 10, 24})
    ->Args({1, 10, 36})
    ->Args({2, 10, 6})
    ->Args({2, 10, 12})
    ->Args({2, 10, 24})
    ->Args({2, 10, 36})
    ->Args({3, 10, 6})
    ->Args({3, 10, 12})
    ->Args({3, 10, 24})
    ->Args({3, 10, 36})
```

**Describe alternatives you've considered**
At the moment, I have a large number of individual `->Args({...})` calls throughout my own benchmarking code, which is quite repetitive and error prone when copy and pasting with slight modifications. Other than that, I am not aware of any other way to achieve the desired functionality.

**Additional context**
I have already implemented this feature on my fork of the project and would like to merge it into the main repository if there are no objections against the proposition.

**Comment by dominichamon** [comment]
this seems utterly reasonable. I think i'd nitpick the name, if you don't mind. While it is a Cartesian product of vectors, that doesn't explain to the user what it's doing. Perhaps something like `ArgsCombination` or `ArgsProduct` ? i'm terrible at naming, but something that indicates that you're setting the args would be better i think.

**Comment by chriswasser** [comment]
I agree `CartesianProduct` might not convey the intent to the user well enough. Your suggestion `ArgsProduct` sounds really good to me. Luckily, we are not the first ones implementing a cartesian product so I decided to look around for some naming conventions. Although not many languages have a cartesian product function in their standard library, here are some notable implementations:
- In C++, range-v3 introduced `view::cartesian_product` ==> hinting towards `ArgsProduct`
- In Python, `itertools` calls it `product` ==> hinting towards `ArgsProduct`
- In Rust, `itertools` calls it `iproduct` ==> hinting towards `ArgsProduct`
- In Ruby, arrays offer a `product` function ==> hinting towards `ArgsProduct`
- In R, it is implemented in `expand.grid` ==> hinting towards `ArgsGrid`
- In R, `purrr` calls it `cross` ==> hinting towards `ArgsCross`
- In SQL, `CARTESIAN JOIN` / `CROSS JOIN` exist ==> hinting towards `ArgsCross`

From the given options, I prefer either `ArgsProduct` or `ArgsGrid`. As most C like languages - and especially C++ - call this functionality a product, I went with `ArgsProduct`. See #1029 for the new commit.
